api4/team.go

```diff
@@ -1193,7 +1206,45 @@ func inviteUsersToTeam(c *Context, w http.ResponseWriter, r *http.Request) {
 	auditRec.AddMeta("emails", emailList)
 
 	if graceful {
-		invitesWithError, err := c.App.InviteNewUsersToTeamGracefully(emailList, c.Params.TeamId, c.App.Session().UserId)
+
+		cloudUserLimit := *c.App.Config().ExperimentalSettings.CloudUserLimit
+
+		// If the cloudUserLimit is > 0, then we have to check the user limit
+		var invitesNotSent []*model.EmailInviteWithError
+		if cloudUserLimit > 0 && c.IsSystemAdmin() {
+			systemUserCount, err := c.App.Srv().Store.User().Count(model.UserCountOptions{})
+			if err != nil {
+				return
+			}
+			remainingUsers := cloudUserLimit - systemUserCount
+			if remainingUsers <= 0 {
+				// No remaining users so all fail
+				invitesNotSent = genEmailInviteWithErrorList(emailList)
+				emailList = nil
+			} else if remainingUsers < int64(len(emailList)) {
+				// Trim the email list to only invite as many users as are remaining in subscription
+				// Set graceful errors for the remaining email addresses
+				emailsAboveLimit := emailList[remainingUsers:]
+				invitesNotSent = genEmailInviteWithErrorList(emailsAboveLimit)
+				// If 1 user remaining we have to prevent 0:0 reslicing
+				if remainingUsers == 1 {
+					email := emailList[0]
+					emailList = nil
+					emailList = append(emailList, email)
+				} else {
+					emailList = emailList[:(remainingUsers - 1)]
+				}
+			}
+		}
+		var invitesWithError []*model.EmailInviteWithError
+		var err *model.AppError
+		if emailList != nil {
+			invitesWithError, err = c.App.InviteNewUsersToTeamGracefully(emailList, c.Params.TeamId, c.App.Session().UserId)
+		}
+		// Add the EmailInviteWithError's we have from user's over the limit
+		if len(invitesNotSent) > 0 {
+			invitesWithError = append(invitesWithError, invitesNotSent...)
+		}
```

Comment:
Generally we try to keep the logic inside the api handlers to a minimum and we avoid any calls to the database in them. Instead, we use the `app` package to handle all that. With that in mind, can you factor this out as a function into the `app` package?

Reply:
I've shuffled around the meatier part of the logic into the app package.